File: scripts/build-hsk-packs.py
#!/usr/bin/env python3
"""
Convert drkameleon/complete-hsk-vocabulary JSON into taijobi pack format.
Generates one pack per HSK level (1-6 old, 1-9 new).

Usage:
  curl -sL https://raw.githubusercontent.com/drkameleon/complete-hsk-vocabulary/main/complete.json -o /tmp/hsk-complete.json
  python3 scripts/build-hsk-packs.py /tmp/hsk-complete.json taijobi-web/static/packs/
"""

import json
import os
import sys


def convert_level(level_str: str) -> tuple[str, int]:
    """Convert 'old-3' or 'new-4' to (system, level_num)."""
    parts = level_str.split('-')
    return parts[0], int(parts[1])


def main():
    if len(sys.argv) < 3:
        print(f"Usage: {sys.argv[0]} <complete.json> <output_dir>")
        sys.exit(1)

    with open(sys.argv[1], 'r', encoding='utf-8') as f:
        data = json.load(f)

    output_dir = sys.argv[2]
    os.makedirs(output_dir, exist_ok=True)

    # Group by old HSK levels (1-6) for now
    levels: dict[int, list] = {}
    for entry in data:
        for level_str in entry.get('level', []):
            system, num = convert_level(level_str)
            if system == 'old' and 1 <= num <= 6:
                if num not in levels:
                    levels[num] = []
                form = entry['forms'][0] if entry.get('forms') else None
                if not form:
                    continue
                word = {
                    'word': entry['simplified'],
                    'pinyin': form['transcriptions'].get('pinyin', ''),
                    'translation': '; '.join(form.get('meanings', [])),
                }
                levels[num].append(word)

    for level_num, words in sorted(levels.items()):
        pack_id = f'hsk-{level_num}'
        # Split into lessons of ~25 words each
        lesson_size = 25
        lessons = []
        for i in range(0, len(words), lesson_size):
            chunk = words[i:i + lesson_size]
            lesson_num = (i // lesson_size) + 1
            lessons.append({
                'id': f'{pack_id}-{lesson_num:02d}',
                'title': f'HSK {level_num} - Teil {lesson_num}',
                'sort_order': lesson_num,
                'vocabulary': chunk,
            })

        pack = {
            'id': pack_id,
            'name': f'HSK {level_num}',
            'version': 1,
            'language_pair': 'zh-en',
            'lessons': lessons,
        }

        out_path = os.path.join(output_dir, f'{pack_id}.json')
        with open(out_path, 'w', encoding='utf-8') as f:
            json.dump(pack, f, ensure_ascii=False, indent=2)

        total_words = sum(len(l['vocabulary']) for l in lessons)
        print(f"  {pack_id}: {total_words} words, {len(lessons)} lessons -> {out_path}")

    print(f"Done: {len(levels)} packs")


if __name__ == '__main__':
    main()

File: scripts/build-hsk-packs.js
#!/usr/bin/env bun
/**
 * Convert drkameleon/complete-hsk-vocabulary JSON into taijobi pack format.
 * Generates one pack per HSK level (1-6 old).
 *
 * Usage:
 *   curl -sL https://raw.githubusercontent.com/drkameleon/complete-hsk-vocabulary/main/complete.json -o /tmp/hsk-complete.json
 *   bun scripts/build-hsk-packs.js /tmp/hsk-complete.json taijobi-web/static/packs/
 */

import { readFileSync, writeFileSync, mkdirSync } from "node:fs";
import { join } from "node:path";

function convertLevel(levelStr) {
  const parts = levelStr.split("-");
  return { system: parts[0], num: parseInt(parts[1], 10) };
}

function main() {
  if (process.argv.length < 4) {
    console.error(`Usage: ${process.argv[1]} <complete.json> <output_dir>`);
    process.exit(1);
  }

  const data = JSON.parse(readFileSync(process.argv[2], "utf-8"));
  const outputDir = process.argv[3];
  mkdirSync(outputDir, { recursive: true });

  // Group by old HSK levels (1-6)
  const levels = new Map();
  for (const entry of data) {
    for (const levelStr of entry.level || []) {
      const { system, num } = convertLevel(levelStr);
      if (system === "old" && num >= 1 && num <= 6) {
        if (!levels.has(num)) levels.set(num, []);
        const form = entry.forms?.[0];
        if (!form) continue;
        levels.get(num).push({
          word: entry.simplified,
          pinyin: form.transcriptions?.pinyin || "",
          translation: (form.meanings || []).join("; "),
        });
      }
    }
  }

  const sortedKeys = [...levels.keys()].sort((a, b) => a - b);
  for (const levelNum of sortedKeys) {
    const words = levels.get(levelNum);
    const packId = `hsk-${levelNum}`;

    // Split into lessons of ~25 words each
    const lessonSize = 25;
    const lessons = [];
    for (let i = 0; i < words.length; i += lessonSize) {
      const chunk = words.slice(i, i + lessonSize);
      const lessonNum = Math.floor(i / lessonSize) + 1;
      lessons.push({
        id: `${packId}-${String(lessonNum).padStart(2, "0")}`,
        title: `HSK ${levelNum} - Teil ${lessonNum}`,
        sort_order: lessonNum,
        vocabulary: chunk,
      });
    }

    const pack = {
      id: packId,
      name: `HSK ${levelNum}`,
      version: 1,
      language_pair: "zh-en",
      lessons,
    };

    const outPath = join(outputDir, `${packId}.json`);
    writeFileSync(outPath, JSON.stringify(pack, null, 2) + "\n", "utf-8");

    const totalWords = lessons.reduce((s, l) => s + l.vocabulary.length, 0);
    console.log(
      `  ${packId}: ${totalWords} words, ${lessons.length} lessons -> ${outPath}`,
    );
  }

  console.log(`Done: ${levels.size} packs`);
}

main();
